File: bench/binarytrees/binarytrees.php
<?
/* 
   The Computer Language Benchmarks Game
   http://shootout.alioth.debian.org/

   contributed by Peter Baltruschat
   multicore by anon
 */

function bottomUpTree($item, $depth)
{
   if($depth)
   {
      --$depth;
      $newItem = $item<<1;
      return array(
         bottomUpTree($newItem - 1, $depth),
         bottomUpTree($newItem, $depth),
         $item
      );
   }
   return array(NULL, NULL, $item);
}

function itemCheck($treeNode)
{
   $check = 0;
   do {
      $check += $treeNode[2];
      if(NULL == $treeNode[0])
      {
         return $check;
      }
      $check -= itemCheck($treeNode[1]);
      $treeNode = $treeNode[0];
   }
   while(TRUE);
}

function getProcs() {
   $procs = 1;
   if (file_exists('/proc/cpuinfo')) {
      $procs = preg_match_all('/^processor\s/m', file_get_contents('/proc/cpuinfo'), $discard);
   }
   $procs <<= 1;
   return $procs;
}

$minDepth = 4;

$n = ($argc == 2) ? $argv[1] : 1;
$maxDepth = max($minDepth + 2, $n);
$stretchDepth = $maxDepth + 1;

$stretchTree = bottomUpTree(0, $stretchDepth);
printf("stretch tree of depth %d\t check: %d\n",
   $stretchDepth, itemCheck($stretchTree));
unset($stretchTree);

$longLivedTree = bottomUpTree(0, $maxDepth);

// # of processes to run
$procs = getProcs();
if ($procs > (($maxDepth - $minDepth + 2) >> 1)) {
   $procs = ($maxDepth - $minDepth + 2) >> 1;
}

// queue jobs
$tok = ftok(__FILE__, chr(time() & 255));
$queue = msg_get_queue($tok);
for ($d = $minDepth; $d <= $maxDepth; $d+=2) {
   $errcode = 0;
   if (!msg_send($queue, 1, (string) $d, FALSE, FALSE, $errcode)) {
      if ($errcode == MSG_EAGAIN) {
         echo "EAGAIN\n";
      }
      var_dump(msg_stat_queue($queue));
      exit(1);
   }
}

// create processes
$parent = TRUE;
for ($i = 1; $i < $procs; ++$i) {
   $pid = pcntl_fork();
   if ($pid === -1) {
      die('could not fork');
   } else if ($pid) {
      continue;
   }
   $parent = FALSE;
   break;
}

// run jobs
ob_start();

$d = 0;
while (msg_receive($queue, 1, $msgtype, 16, $d, FALSE, MSG_IPC_NOWAIT)) {
   $d = (int) $d;
   $iterations = 1 << ($maxDepth - $d + $minDepth);
   $check = 0;
   for($i = 1; $i <= $iterations; ++$i)
   {
      $t = bottomUpTree($i, $d);
      $check += itemCheck($t);
      unset($t);
      $t = bottomUpTree(-$i, $d);
      $check += itemCheck($t);
      unset($t);
   }

   printf("%d\t trees of depth %d\t check: %d\n",
      $iterations<<1, $d, $check);

}

if (!$parent) {
   msg_send($queue, 2, ob_get_clean(), FALSE);
   exit(0);
}

// gather & output results
$s = 0;
$output = array(explode("\n", ob_get_clean()));
for ($i = 1; $i < $procs; ++$i) {
   msg_receive($queue, 2, $msgtype, 4096, $o, FALSE);
   $output[] = explode("\n", $o);
   pcntl_wait($s);
}

msg_remove_queue($queue);

$output = call_user_func_array('array_merge', $output);
$output = array_filter($output, 'strlen');
rsort($output, SORT_NUMERIC);
echo implode("\n", $output), "\n";

printf("long lived tree of depth %d\t check: %d\n",
   $maxDepth, itemCheck($longLivedTree));

File: bench/binarytrees/binarytrees.scala
/* The Computer Language Shootout
   http://shootout.alioth.debian.org/

   contributed by Kannan Goundan
   modified by Isaac Gouy
   optimized by David Pollak
   updated to 2.8 by Rex Kerr
   modified by Piotr Tarsa
*/

sealed abstract class Node(i: Int, left: Node, right: Node) {
  def isum: Int
}
case class NonLeaf(i: Int, left: Node, right: Node) extends Node(i, left, right) {
  def isum: Int = i + left.isum - right.isum
}
case class Leaf(i: Int) extends Node(i, NullNode, NullNode) {
  def isum: Int = i
}
case object NullNode extends Node(0, new Leaf(0), new Leaf(0)) {
  def isum: Int = 0
}

object Tree {
  def apply(i: Int, depth: Int): Node = {
    if (depth > 0) NonLeaf(i, Tree(i * 2 - 1, depth - 1), Tree(i * 2, depth - 1))
    else Leaf(i)
  }
}

object binarytrees {
  def main(args: Array[String]) = {
    val n = try{ args(0).toInt } catch { case _ => 1 }
    val minDepth = 4
    val maxDepth = n max (minDepth + 2)

    def print(name: String, depth: Int, check: Int) =
      println(name + " of depth " + depth + "\t check: " + check)

    print("stretch tree", maxDepth + 1, Tree(0, maxDepth + 1).isum)
    val longLivedTree = Tree(0, maxDepth)
    minDepth to maxDepth by 2 foreach {
      depth =>
      val iterations = 1 << (maxDepth - depth + minDepth)
      var i, sum = 0
      while (i < iterations) {
        i += 1
        sum += Tree(i, depth).isum + Tree(-i, depth).isum
      }
      print(iterations *2  + "\t trees", depth, sum)
    }
    print("long lived tree", maxDepth, longLivedTree.isum)
  }
}